### PR TITLE
`Link.serialize` to support ChainerX

### PR DESCRIPTION
**Pull request by hvy**
Fixes https://github.com/chainer/chainer/issues/7172.

### CODE REVIEW

File: tests/chainer_tests/test_link.py [hunk 1]
```diff
@@ -2466,4 +2466,43 @@ def test_to_intel64(self):
         assert l.child.to_method_called == 1
 
 
+class TestSerialize(unittest.TestCase):
+    def setUp(self):
+        self.array = numpy.array([1, 2, 3], dtype=numpy.float32)
+        self.serializer = mock.MagicMock(return_value=self.array)
+
+        # Create an empty link to populate.
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Is this comment appropriate? `link` is not empty a few lines below.

File: tests/chainer_tests/test_link.py [hunk 2]
```diff
@@ -2466,4 +2466,43 @@ def test_to_intel64(self):
...
+        link.to_device(device_spec)
+        link.serialize(serializer)
+
+        self.assertEqual(serializer.call_count, 3)
+        cpu_device = chainer.backend.CpuDevice()
+        numpy.testing.assert_array_equal(cpu_device.send(link.x.array), array)
+        numpy.testing.assert_array_equal(cpu_device.send(link.y.array), array)
+        numpy.testing.assert_array_equal(cpu_device.send(link.z), array)
+
+    def test_serialize_numpy(self):
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Why not use `inject_backend_config`?

File: tests/chainer_tests/test_link.py (hunk 1 above)
Comment (hvy) [reply tests/chainer_tests/test_link.py]: Removed the comment.

File: tests/chainer_tests/test_link.py (hunk 2 above)
Comment (hvy) [reply tests/chainer_tests/test_link.py]: Thanks, that actually helped me find a bug when the device is a ChainerX CUDA device. Fixed that bug and the test.

**Comment by hvy** [comment]
Thanks for the comments. PTAL.

**Comment by niboshi** [comment]
Jenkins, test this please

**Pull request merged**